spring-cloud-dataflow-completion/src/main/java/org/springframework/cloud/dataflow/completion/ExpansionStrategy.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.dataflow.completion;
+
+import java.util.List;
+
+import org.springframework.cloud.dataflow.core.StreamDefinition;
+
+/**
+ * Used to enhance a well formed stream definition by adding yet more text to it (<i>e.g.</i> adding more options to a
+ * module).
+ * 
+ * @author Eric Bottard
+ */
+public interface ExpansionStrategy {
```

Comment:
This interface requires two invocations - one to determine if the strategy applies, another to mutate a list passed to it.

A few observations about this interface:

1. It requires the caller to invoke `shouldTrigger` to determine if the strategy applies to the parse result. What happens if the caller does not invoke this method but invokes `addProposals` instead?

2. The `addProposals` method returns `void` and mutates a collection passed to it. It assumes that the collection being passed to it is mutable.

Could we combine these aspects into a single method for this interface?

```
/**
* For a given stream DSL text and {@link StreamDefinition}, return a list
* of {@link CompletionProposal} that apply.
* 
* @param text DSL text for the stream
* @param streamDefinition stream definition
* @param detailLevel integer representing the amount of detail to include
* in the generated {@code CompletionProposal}s
* @return a list of {@code CompletionProposal}; if no proposals apply
* an empty list is returned
*/
List<CompletionProposal> generateProposals(String text, StreamDefinition streamDefinition, int detailLevel);
```
This way, callers can maintain their own `List` of proposals, and the invocation to `shouldTrigger` won't be required because an empty `List` will be returned if the completion should not be triggered.